src/Conversion/ONNXToTOSA/Tensor/Resize.cpp

```diff
@@ -0,0 +1,213 @@
+  /// - scale_x_d = input_size
+  /// - scale_x_n = output_size
+  /// - offset = 0
+  ///
+  /// ## nearest_mode ##
+  /// If mode == nearest, then ONNX can set the nearest_mode attribute.
+  /// The standard case for TOSA of round_half_up. Support for floor can
+  /// be achieved with:
+  /// offset_x -= n/2
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    auto resizeOp = llvm::cast<ONNXResizeOp>(op);
+    Location loc = op->getLoc();
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+
+    Value input = adaptor.X();
+    auto inputType = input.getType().dyn_cast<RankedTensorType>();
+
+    auto resultType =
+        resizeOp.getResult().getType().dyn_cast<RankedTensorType>();
+
+    StringRef mode = adaptor.mode();
+    StringRef nearestMode = adaptor.nearest_mode();
+    StringRef coordinateTransformationMode =
+        adaptor.coordinate_transformation_mode();
+
+    if (inputType.getRank() != 4) {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA only support 4D tensors as input of resize.");
+    }
+
+    if (inputType.isDynamicDim(2) || inputType.isDynamicDim(3)) {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "Only static sized tensors are supported.");
+    }
+
+    if (mode == "cubic") {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA does not support cubic interpolation.");
+    }
+
+    if (mode == "nearest" &&
+        (nearestMode == "ceil" || nearestMode == "round_prefer_floor")) {
+      return rewriter.notifyMatchFailure(resizeOp,
+          "TOSA does not support ceil and round_prefer_floor as nearestMode.");
+    }
+
+    if (coordinateTransformationMode == "tf_crop_and_resize") {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA does not support tf_crop_and_resize.");
+    }
+
+    // Convert input [N,IC,IH,IW] -> [N,IH,IW,IC]
+    Value newInput = tosa::createTosaTransposedTensor(
+        rewriter, resizeOp, input, {0, 2, 3, 1});
+
+    bool alignCorners = coordinateTransformationMode == "align_corners";
+    bool halfPixel = coordinateTransformationMode == "half_pixel";
+    bool pytorchHalfPixel =
+        coordinateTransformationMode == "pytorch_half_pixel";
+    bool isBilinear = mode == "linear";
+    bool isNearest = mode == "nearest";
+    bool floor = nearestMode == "floor";
+    StringRef resizeMode = isBilinear ? "BILINEAR" : "NEAREST_NEIGHBOR";
+
+    auto inputShape = inputType.getShape();
+    auto outputShape = resultType.getShape();
+
+    int64_t inputHeight = inputShape[2];
+    int64_t inputWidth = inputShape[3];
+    int64_t outputHeight = outputShape[2];
+    int64_t outputWidth = outputShape[3];
+
+    // Adapted from TFL to TOSA.
+    // Align corners sets the scaling ratio to (OH - 1)/(IH - 1)
+    // rather than OH / IH. Similarly for width.
+    auto normalize = [&](int64_t input, int64_t output, int64_t &n, int64_t &d,
```

Comment:
I wonder if it wouldn't be more readable to have `n`, `d`, etc. as return values instead of output parameters. Something like `return {n, d, offset, border}` and then have the call to `normalize` as `auto [scale_y_n, scale_y_d, offset_y, border_y] = normalize(...)` ?

Reply:
I like your approach. I will adapt it.